### PR TITLE
feat: replace Jansi with JLine to resolve Java 21+ native access warnings

### LINKED ISSUE #603: Replace Jansi with JLine to resolve Java 21+ native access warnings
## Problem

When running the CLI on Java 21+, the following warnings are displayed:

```
WARNING: A restricted method in java.lang.System has been called
WARNING: java.lang.System::load has been called by org.fusesource.jansi.internal.JansiLoader in an unnamed module
WARNING: Use --enable-native-access=ALL-UNNAMED to avoid a warning for callers in this module
WARNING: Restricted methods will be blocked in a future release unless native access is enabled
```

This is caused by Jansi's use of native code loading via `System.load()`, which Java 21+ restricts as part of [JEP 472: Prepare to Restrict the Use of JNI](https://openjdk.org/jeps/472).

## Current State

- Jansi version: 2.4.2 (`org.fusesource.jansi:jansi`)
- The warning is cosmetic but will become a blocking error in future Java releases
- The `--enable-native-access=ALL-UNNAMED` flag doesn't exist on older JREs, making it difficult to add conditionally

## Proposed Solution

Migrate from Jansi to JLine:

1. **Jansi has been merged into JLine** - As of JLine 3.25, Jansi was merged into the JLine project
2. **JLine 4.x removes Jansi entirely** - Uses a JNI provider that handles native access more gracefully
3. **JLine provides the FFM provider** - For JDK 22+, JLine offers a Foreign Function & Memory (FFM) provider as an alternative to JNI

### Migration Path

- Replace `org.fusesource.jansi:jansi` with `org.jline:jline` or specific JLine modules
- Update `cli-processor` to use JLine's terminal/console APIs
- The `AnsiConsole` and ANSI rendering functionality should have equivalent JLine APIs

## References

- [JLine GitHub](https://github.com/jline/jline3)
- [Maven has the same issue - #11289](https://github.com/apache/maven/issues/11289)
- [JLine JPMS Support](https://jline.org/docs/modules/jpms/)

## Alternatives Considered

1. **Custom launcher scripts** - Add Java version detection to conditionally add `--enable-native-access=ALL-UNNAMED`
2. **Document JAVA_OPTS workaround** - Users manually set the flag
3. **Accept the warning** - Functional but not ideal for user experience

The JLine migration is preferred as it addresses the root cause and prepares for future Java releases.

### PR DESCRIPTION
**Pull request by david-waltermire**
## Summary

Resolves #603 by migrating from `org.fusesource.jansi:jansi` 2.4.2 to `org.jline:jansi-core` 3.30.6 and eliminating native terminal detection entirely.

JLine 3.25+ includes the Jansi codebase directly under the `org.jline.jansi` package. However, simply switching dependencies still triggered native access warnings because `AnsiConsole.systemInstall()` loads native libraries for terminal detection. This PR eliminates native access by relying on modern terminals' built-in ANSI support instead.

### Dependency Migration
- Replace `org.fusesource.jansi:jansi` with `org.jline:jansi-core` 3.30.6 (minimal module for ANSI support, ~245KB vs ~1.8MB for full JLine)
- Update imports from `org.fusesource.jansi` to `org.jline.jansi`

### Native Access Elimination
- **Remove `AnsiConsole.systemInstall()`** - This call triggered native library loading which causes warnings on JDK 22+ (JEP 454) and errors on JDK 24+ (JEP 472)
- **Use `System.out` directly** - Modern terminals (Windows 10+, Linux, macOS) support ANSI escape codes natively without detection
- **Replace `AnsiPrintStream.getTerminalWidth()`** - Now uses `COLUMNS` environment variable with 80-column fallback
- **Use `Ansi.setEnabled(false)`** for `--no-color` mode instead of AnsiConsole properties

### Help Output Improvements
- **Fix text alignment** - Added ANSI-aware text wrapping for command descriptions to prevent misalignment when ANSI codes are present
- **Hide "Since" column** - Removed the empty "Since" column from options table for cleaner output

### Input Validation
- **wrapText() validation** - Added IllegalArgumentException for invalid inputs (maxWidth <= 0, indent >= maxWidth) to prevent infinite loops

### Flaky Test Fix
- **DomDatatypeContent thread safety** - Fixed sporadic `fNodeListCache` null pointer errors during parallel test execution by importing DOM elements into a fresh Document for each instance, ensuring complete DOM tree isolation

### Breaking Changes
None expected. The CLI behavior should be identical.

### Behavioral Differences

| Aspect | Before | After |
|--------|--------|-------|
| Terminal width detection | Native call via JNI/FFM | `COLUMNS` env var or 80 default |
| ANSI support detection | Native terminal query | Assumes ANSI support (modern default) |
| Java 21+ warnings | `WARNING: A restricted method in java.lang.System has been called` | No warnings |

### Terminal Width
The CLI now reads terminal width from the `COLUMNS` environment variable, which is set automatically by most shells (bash, zsh, fish, PowerShell). If not set, it defaults to 80 columns. This may affect help text wrapping in edge cases where:
- The terminal doesn't set `COLUMNS`
- The terminal is resized after the shell started (some shells update `COLUMNS` dynamically, others don't)

### ANSI Color Support
The CLI now assumes the terminal supports ANSI escape codes. This is true for:
- Windows 10 version 1511+ (November 2015) with Windows Terminal or modern cmd.exe
- All modern Linux terminals
- macOS Terminal.app and iTerm2
- VS Code integrated terminal

### Legacy Console Support
Users on legacy consoles that do not support ANSI escape codes (e.g., older Windows cmd.exe, certain CI/CD environments, or redirected output) may see raw escape sequences in the output. To address this, use the `--no-color` flag to disable ANSI formatting:

```bash
metaschema-cli --no-color <command>
```

## Test Plan

- [x] Run `mvn -pl cli-processor,metaschema-cli test` - All CLI tests pass
- [x] Run `mvn clean install -PCI -Prelease` - Full build passes (5078 tests, 0 failures, 0 errors)
- [x] Unit tests for text wrapping functionality (17 tests including edge cases)
- [x] Unit tests for wrapText() input validation (4 tests)
- [x] Verify schemagen tests no longer flaky with DOM isolation fix
- [ ] Manual testing of CLI with and without `--no-color`
- [ ] Verify no native access warnings on Java 21+/24+

## Build Verification

```
Build verified successfully:
- ✅ Tests: 5078 passed, 0 failed, 0 errors
- ✅ SpotBugs: 0 bugs, 0 errors
- ✅ PMD: 0 violations
- ✅ Checkstyle: 0 violations
```

## Summary by CodeRabbit

* **New Features**
  * Added --no-color flag to disable colored terminal output.

* **Improvements**
  * Improved ANSI handling for broader compatibility.
  * Enhanced help formatting with terminal-width-aware wrapping.
  * Made XML/schema generation more robust and thread-safe.

* **Documentation**
  * Added CLI usage notes and distro guidance for disabling color output.

* **Tests**
  * Added tests for version output, no-color behavior, and text-wrapping utilities.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: cli-processor/src/main/java/gov/nist/secauto/metaschema/cli/processor/CallingContext.java [hunk 1]
```diff
@@ -540,6 +547,84 @@ private static CharSequence getExtraArguments(@NonNull ICommand targetCommand) {
...
+      int lastSpace = text.lastIndexOf(' ', lineEnd);
+
+      if (lastSpace <= lineStart) {
+        // No space found, force break at width
+        lastSpace = lineEnd;
+      }
+
+      result.append(text, lineStart, lastSpace);
+      lineStart = lastSpace + 1; // Skip the space
+      firstLine = false;
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply cli-processor/src/main/java/gov/nist/secauto/metaschema/cli/processor/CallingContext.java:633]: @coderabbitai Good catch\! Fixed the off-by-one bug in the force-break logic. The fix separates the handling of force-break (no space found) from word-boundary wrapping to avoid incorrectly skipping a character. Added two new unit tests to verify all characters are preserved during force-break wrapping.

File: pom.xml [hunk 2]
```diff
@@ -51,7 +51,7 @@
 		<dependency.freemarker.version>2.3.34</dependency.freemarker.version>
 		<dependency.inject-resources-junit-jupiter.version>1.0.0</dependency.inject-resources-junit-jupiter.version>
 		<dependency.jackson.version>2.20.1</dependency.jackson.version>
-		<dependency.jansi.version>2.4.2</dependency.jansi.version>
+		<dependency.jline.version>3.28.0</dependency.jline.version>
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply pom.xml]: @coderabbitai Confirmed. JLine 3.28.0 (released Dec 10, 2024) is the latest stable version.

[the remark this answers is not in the document]
Comment (david-waltermire) [reply pom.xml]: @coderabbitai Correction: Updated to JLine 3.30.6 (released Dec 30, 2025) which is the actual latest version.

File: cli-processor/src/main/java/gov/nist/secauto/metaschema/cli/processor/CallingContext.java [hunk 3]
```diff
@@ -421,16 +421,23 @@ private String buildHelpFooter() {
...
-      int length = subCommands.stream()
+      int commandColWidth = subCommands.stream()
           .mapToInt(command -> command.getName().length())
           .max().orElse(0);
 
+      // Calculate description column width: terminal - 3 (leading spaces) -
+      // commandCol - 1 (space)
+      int prefixWidth = 3 + commandColWidth + 1;
+      int descWidth = Math.max(terminalWidth - prefixWidth, 20);
+      String continuationIndent = " ".repeat(prefixWidth);
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply cli-processor/src/main/java/gov/nist/secauto/metaschema/cli/processor/CallingContext.java:432]: @coderabbitai Addressed in commit c34738de4. Added input validation to wrapText() that throws IllegalArgumentException when maxWidth <= 0 or indent.length() >= maxWidth, preventing infinite loops.

File: cli-processor/src/main/java/gov/nist/secauto/metaschema/cli/processor/CallingContext.java [hunk 4]
```diff
@@ -540,6 +547,85 @@ private static CharSequence getExtraArguments(@NonNull ICommand targetCommand) {
...
+      if (lastSpace <= lineStart) {
+        // No space found, force break at width
+        result.append(text, lineStart, lineEnd);
+        lineStart = lineEnd; // Continue from break point (no space to skip)
+      } else {
+        result.append(text, lineStart, lastSpace);
+        lineStart = lastSpace + 1; // Skip the space
+      }
+      firstLine = false;
+    }
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply cli-processor/src/main/java/gov/nist/secauto/metaschema/cli/processor/CallingContext.java:634]: @coderabbitai Addressed in commit c34738de4. Added input validation at the start of wrapText() that throws IllegalArgumentException for invalid parameters, preventing the potential infinite loop.

File: cli-processor/src/main/java/gov/nist/secauto/metaschema/cli/processor/CLIProcessor.java [hunk 5]
```diff
@@ -273,8 +270,8 @@ protected final Map<String, ICommand> getTopLevelCommandsByName() {
   }
 
   static void handleNoColor() {
-    System.setProperty(AnsiConsole.JANSI_MODE, AnsiConsole.JANSI_MODE_STRIP);
-    AnsiConsole.systemUninstall();
+    // Disable ANSI escape sequences - the Ansi class will output plain text
+    Ansi.setEnabled(false);
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply cli-processor/src/main/java/gov/nist/secauto/metaschema/cli/processor/CLIProcessor.java:281]: @coderabbitai Addressed in commit c34738de4. Added Javadoc to handleNoColor() documenting its purpose and usage scenarios.

File: cli-processor/src/test/java/gov/nist/secauto/metaschema/cli/processor/CallingContextTest.java [hunk 6]
```diff
@@ -266,4 +266,123 @@ void appliesBothQuietAndNoColor() throws ParseException {
...
+      String text = "ABCDEFGHIJKLMNO"; // 15 chars
+      String result = CallingContext.wrapText(text, 5, "");
+
+      String[] lines = result.split(System.lineSeparator());
+      assertEquals(3, lines.length, "Should produce 3 lines of 5 chars each");
+      assertEquals("ABCDE", lines[0]);
+      assertEquals("FGHIJ", lines[1]);
+      assertEquals("KLMNO", lines[2]);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply cli-processor/src/test/java/gov/nist/secauto/metaschema/cli/processor/CallingContextTest.java:428]: @coderabbitai Addressed in commit c34738de4. Added edge case tests for wrapText() validation: throws IllegalArgumentException when maxWidth is zero/negative or when indent length >= maxWidth. Also added a test for very narrow width with word boundaries.

File: cli-processor/src/main/java/gov/nist/secauto/metaschema/cli/processor/CallingContext.java [hunk 7]
```diff
@@ -540,6 +547,92 @@ private static CharSequence getExtraArguments(@NonNull ICommand targetCommand) {
...
+   */
+  @NonNull
+  static String wrapText(@NonNull String text, int maxWidth, @NonNull String indent) {
+    if (maxWidth <= 0) {
+      throw new IllegalArgumentException("maxWidth must be positive, got: " + maxWidth);
+    }
+    if (indent.length() >= maxWidth) {
+      throw new IllegalArgumentException(
+          "indent length (" + indent.length() + ") must be less than maxWidth (" + maxWidth + ")");
+    }
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply cli-processor/src/main/java/gov/nist/secauto/metaschema/cli/processor/CallingContext.java:599]: @coderabbitai Addressed in commit 7854810c3. Added @throws IllegalArgumentException tag documenting the validation cases for maxWidth and indent length.

File: schemagen/src/main/java/gov/nist/secauto/metaschema/schemagen/xml/impl/DomDatatypeContent.java [hunk 8]
```diff
@@ -44,12 +57,47 @@ public class DomDatatypeContent
...
+      DocumentBuilder builder = factory.newDocumentBuilder();
+      Document doc = builder.newDocument();
+
+      return elements.stream()
+          .map(e -> (Element) doc.importNode(e, true))
+          .collect(Collectors.toList());
+    } catch (ParserConfigurationException ex) {
+      throw new IllegalStateException("Failed to create DocumentBuilder for DOM isolation", ex);
+    }
   }
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply schemagen/src/main/java/gov/nist/secauto/metaschema/schemagen/xml/impl/DomDatatypeContent.java:105]: @coderabbitai Addressed in commit 7854810c3. Added @throws IllegalStateException tag documenting the DocumentBuilder creation failure case.

**Comment by david-waltermire** [comment]
@coderabbitai Addressed all nitpick suggestions in commit 3ba10a15a:

1. **LoggingValidationHandlerTest**: Added explicit `isNotSameAs()` assertion to verify that `instance(true)` and `instance(false)` return different singleton instances.

2. **CLIProcessorTest**: Refactored version output tests to use JUnit 5 parameterized tests with `@CsvSource`, reducing duplication while maintaining clear test output and failure messages.

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/model/metaschema/binding/InlineDefineField.java [hunk 9]
```diff
@@ -4,6 +4,7 @@
  */
 // Generated from: ../../../../../../../../../../../../core/metaschema/schema/metaschema/metaschema-module-metaschema.xml
 // Do not edit - changes will be lost when regenerated.
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply databind/src/main/java/gov/nist/secauto/metaschema/databind/model/metaschema/binding/InlineDefineField.java:6]: right. These were regenerated.

**Comment by david-waltermire** [comment]
@coderabbitai Addressed additional nitpick suggestions in commits 3ba10a15a and df6f0f718:

1. **LoggingValidationHandlerTest** - Added test verifying `instance()` and `instance(false)` return the same singleton, strengthening contract verification.

2. **CLIProcessorTest** - Refactored version output tests to use JUnit 5 parameterized tests.

3. **DomDatatypeContent constructor** - Added `@throws IllegalStateException` to constructor Javadoc per coding guidelines.

Build verified:
- ✅ Tests: All 11 tests pass in LoggingValidationHandlerTest
- ✅ SpotBugs: 0 bugs, 0 errors
- ✅ PMD: 0 violations
- ✅ Checkstyle: 0 violations
- ✅ Jacoco: All coverage checks passed
